capstone/src/test/java/com/google/sps/clubs/ClubServletTest.java

```diff
@@ -65,70 +69,114 @@ public void setUp() throws IOException {
     this.blobstoreServlet = new BlobstoreServlet();
     datastore = DatastoreServiceFactory.getDatastoreService();
     blobstore = Mockito.mock(BlobstoreService.class);
-  }
-
-  @After
-  public void tearDown() {
-    helper.tearDown();
-  }
 
-  @Test
-  public void doPost_registerNewValidClub() throws ServletException, IOException {
     helper.setEnvEmail(TEST_EMAIL).setEnvAuthDomain("google.com").setEnvIsLoggedIn(true);
-    when(request.getParameter(Constants.CLUB_NAME_PROP)).thenReturn(SAMPLE_CLUB_NAME);
+    when(request.getParameter(Constants.PROPERTY_NAME)).thenReturn(SAMPLE_CLUB_NAME);
     when(request.getParameter(Constants.DESCRIP_PROP)).thenReturn(SAMPLE_CLUB_DESC_1);
     when(request.getParameter(Constants.WEBSITE_PROP)).thenReturn(SAMPLE_CLUB_WEB);
 
     ImmutableList<BlobKey> keys = ImmutableList.of(SAMPLE_BLOB);
     ImmutableMap<String, List<BlobKey>> blobMap = ImmutableMap.of(Constants.LOGO_PROP, keys);
     when(blobstore.getUploads(request)).thenReturn(blobMap);
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
 
+  @Test
+  public void doPost_registerNewValidClub() throws ServletException, IOException {
     clubServlet.doPostHelper(request, response, blobstore, datastore);
```

Comment:
Sorry if I didn't catch this last time, but helper methods shouldn't be public. Instead, we should make them private and test doPost instead.

Reply:
So this was actually a problem I encountered when writing first writing the tests. The reason I added a helper method was so I could pass in the mock blobstore. When I tested the doPost directly, the code inside of the doPost uses the real Blobstore API instead. I found that this was a common problem with other STEP projects ([this example](https://github.com/googleinterns/step132-2020/blob/d459dcd86162dcfabc1a1ff9bef5bacfb9cfc363/src/test/java/com/google/sps/RegistrationTest.java) on line 92 is what I referred to while writing this test. Another alternative to this is that I could make the blobstore variable a public or package-private field, so that I could access it from the a test file, although I'm not sure if that is a very good solution. Is there another way to do this/which solution should I go with?

Reply:
Ah I see. I think generally (when not using servlets), you would bind services to the class to define them as sort of a global field in the constructor of the class code. But I'm not entirely sure you can do it with servlets since it seems the accepted pattern is to define the service inside the doPost. This looks good then, thanks for looking into this!

Reply:
Okay, that makes sense. Thank you!